rest-api/dao/participant_counts_over_time_service.py

```diff
@@ -0,0 +1,156 @@
+             DATE(CASE WHEN enrollment_status = 3 THEN
+                   GREATEST(consent_for_electronic_health_records,
+                            questionnaire_on_the_basics_time,
+                            questionnaire_on_lifestyle_time,
+                            questionnaire_on_overall_health_time,
+                            physical_measurements_time,
+                            CASE WHEN sample_status_1ed04_time IS NOT NULL
+                             THEN
+                             (CASE WHEN sample_status_1sal_time IS NOT
+                                NULL
+                                THEN LEAST(sample_status_1ed04_time,
+                                           sample_status_1sal_time)
+                                ELSE sample_status_1ed04_time END)
+                             ELSE sample_status_1sal_time END)
+                   ELSE NULL END) day
+               FROM participant_summary ps
+              WHERE %(filters)s
+           GROUP BY day) full
+             ON c2.day = full.day) day_sums, calendar
+          WHERE calendar.day >= :start_date
+            AND calendar.day <= :end_date
+          GROUP BY calendar.day
+          ORDER BY calendar.day;
+      """ % {'filters': filters_sql}
+
+    params = {'start_date': start_date, 'end_date': end_date}
+
+    results_by_date = []
+
+    with self.session() as session:
+      cursor = session.execute(sql, params)
+
+    # Iterate through each result (by date), transforming tabular SQL results
+    # into expected list-of-dictionaries response format
+    try:
+      results = cursor.fetchall()
+      for result in results:
+        date = result[-1]
+        metrics = {}
+        values = result[:-1]
+        for i, value in enumerate(values):
+          key = strata[i]
+          if value == None:
+            value = 0
+          metrics[key] = int(value)
+        results_by_date.append({
+          'date': str(date),
+          'metrics': metrics
+        })
+    finally:
+      cursor.close()
+
+    return results_by_date
+
+
+  def get_facets_sql(self, facets):
+    """Helper function to transform facets/filters selection into SQL
+
+    :param facets: Object representing facets and filters to apply to query results
+    :return: SQL for 'WHERE' clause, reflecting filters specified in UI
+    """
+
+    facets_sql = []
+
+    facet_map = {
+      'awardee_ids': 'hpo_id',
+      'enrollment_statuses': 'enrollment_status'
+    }
+
+    for facet in facets:
+      filters_sql = []
+      db_field = facet_map[facet]
+      filters = facets[facet]
+      for q_filter in filters:
+        if str(q_filter) != '':
+          filters_sql.append('ps.' + db_field + ' = ' + str(int(q_filter)))
+      if len(filters_sql) > 0:
+        filters_sql = ' OR '.join(filters_sql)
+        facets_sql.append(filters_sql)
+
+    if len(facets_sql) > 0:
+      facets_sql = ' AND '.join(facets_sql)
+    else:
+      facets_sql = '1 = 1'
```

Comment:
Seems like we could return None here and just not include facets_sql when it's None in the main method?

I have a pet peeve about 1=1 in my SQL statements. :)

Reply:
This no longer uses 1=1, and instead inlines the WHERE keyword to this method and returns an empty string when there's no facets_sql.